packages/ipfs-core/src/components/pin/remote.js

```diff
@@ -0,0 +1,441 @@
+      throw new Error('service name must be passed')
+    }
+    const svc = this.service.serviceNamed(service)
+    return svc.rmAll(rmOpts)
+  }
+}
+
+/**
+ * RemotePinServiceAPI provides methods to add, remove, and list the configured
+ * remote pinning services that are used by the remote pinning api.
+ */
+class PinRemoteServiceAPI {
+  constructor ({ config, swarm, peerId }) {
+    this.config = config
+    this.swarm = swarm
+    this.peerId = peerId
+
+    // TODO: read service config from IPFS config to construct remote service at init
+    this._services = {}
+  }
+
+  /**
+   * Adds a new remote pinning service to the set of configured services.
+   *
+   * @param {string} name - the name of the pinning service. Used to identify the service in future remote pinning API calls.
+   * @param {Object} options
+   * @param {string|URL} options.endpoint - the remote API endpoint URL
+   * @param {string} options.key - an API key that authorizes use of the remote pinning service
+   */
+  async add (name, options) {
+    if (this._services[name]) {
+      throw new Error('service already present: ' + name)
+    }
+
+    if (!options.endpoint) {
+      throw new Error('option "endpoint" is required')
+    }
+
+    if (!options.key) {
+      throw new Error('option "key" is required')
+    }
+
+    const svcOpts = Object.assign({ swarm: this.swarm, peerId: this.peerId }, options)
+    this._services[name] = new RemotePinningService(name, svcOpts)
+  }
+
+  /**
+   * List the configured remote pinning services.
+   *
+   * @typedef {object} PinCounts
+   * @property {number} queued
+   * @property {number} pinning
+   * @property {number} pinned
+   * @property {number} failed
+   *
+   * @typedef {object} RemotePinningServiceDescription
+   * @property {string} name
+   * @property {URL} endpoint
+   * @property {?object} stat
+   * @property {PinServiceStatus} stat.status
+   * @property {PinCounts} stat.pinCount
+   *
+   * @param {object} opts
+   * @param {?boolean} opts.stat - if true, include status info for each pinning service
+   * @returns {Promise<Array<RemotePinningServiceDescription>>}
+   */
+  async ls (opts) {
+    const { stat } = (opts || {})
+
+    const promises = []
+    for (const name of Object.keys(this._services)) {
+      const svc = this._services[name]
+      promises.push(svc.info(stat))
+    }
+    return Promise.all(promises)
+  }
+
+  /**
+   * Remove a remote pinning service from the set of configured services.
+   *
+   * @param {string} name - the name of the pinning service to remove
+   * @returns {Promise<void>}
+   */
+  async rm (name) {
+    if (!name) {
+      throw new Error('parameter "name" is required')
+    }
+    delete this._services[name]
+  }
+
+  /**
+   * Returns a RemotePinningService object for the given service name. Throws if no service has been configured with the given name.
+   *
+   * @param {string} name
+   * @returns {RemotePinningService}
+   */
+  serviceNamed (name) {
```

Comment:
I think we would want to make it a static method instead to avoid polluting a namespace.